javascript/javascript3/week3/Shopping Cart/shoppingCart.js

```diff
@@ -0,0 +1,135 @@
+const reducer = (acc, currentVal) => acc + currentVal;
+const apiKey = "8df58438c8513b601e88d88777ea74e8";
+
+const arrayOfProducts = [];
+
+const listItem1 = document.getElementById("list-item-1");
+const listItem2 = document.getElementById("list-item-2");
+
+const userNamePlace = document.getElementById("user-name-spot");
+const pricePlace = document.getElementById("total-price");
+
+class Product {
+  constructor(name, price) {
+    this.name = name;
+    this.price = price;
+  }
+
+  convertToCurrency(currency) {
+    //base currency is euro
+    const currencyConventerAPI = `http://data.fixer.io/api/latest?access_key=${apiKey}`;
+    fetch(currencyConventerAPI)
+      .then(res => res.json())
+      .then(data =>
+        console.log(
+          `${this.name} is ${Math.round(
+            this.price * data.rates[currency]
+          )} ${currency}`
+        )
+      );
+  }
+}
+
+const gum = new Product("gum", 2);
+gum.convertToCurrency("DKK");
+gum.convertToCurrency("USD");
+gum.convertToCurrency("ISK");
+
+class ShoppingCart {
+  constructor(products) {
+    this.products = products;
+  }
+
+  addProduct(product) {
+    this.products.push(product);
+  }
+
+  removeProduct(product) {
+    this.products.forEach(item => {
+      if (this.products.indexOf(product) !== -1) {
+        const filtered = this.products.splice(
+          this.products.indexOf(product),
+          1
+        );
+        return filtered;
+      }
+    });
+  }
+
+  searchProduct(productName) {
+    const product = [];
+    this.products.forEach(item => {
+      if (item.name.includes(productName)) {
+        return product.push(item);
+      }
+    });
+    console.log(product);
+    return product;
+  }
+
+  getTotal() {
+    const productPrices = [];
+    this.products.forEach(item => productPrices.push(item.price));
+    const sum = productPrices.reduce(reducer);
+    pricePlace.innerText = `Total: ${sum}`;
+    sum;
+    return sum;
+  }
+
+  renderProducts() {
+    for (let i = 0; i < this.products.length; i++) {
+      const product = this.products[i];
+      listItem1.appendChild(document.createElement("td")).innerText =
+        product.name;
+      listItem2.appendChild(document.createElement("td")).innerText =
+        product.price;
+    }
+  }
+
+  getUser() {
+    const url = "https://jsonplaceholder.typicode.com/users/1";
+    return new Promise(() => {
```

Comment:
no need for a "new Promise" here. 90% of the time using a "new Promise" is avoidable, especially if you're calling an external API

Reply:
I put it because of getUser & renderProducts exercise, If I don't put fetch into a promise it doesn't get rendered in the page. Do you know in which other ways I could handle this task without returning a promise? Thank you for your comments!

Reply:
I haven't tested it but i think this would work:
``` getUser() {
    const url = "https://jsonplaceholder.typicode.com/users/1";
    return fetch(url)
        .then(res => res.json())
        .then(data => {
          return data.name;
       ///   userNamePlace.innerText = userName; // this belongs in a render function
        });
    });```

and then, since you are returning the userName, you can use it in the caller code:
```shoppingCart.getUser().then(userName => shoppingCart.renderProducts(userName));``` or something like that. This last part is not very precise, it's just to give you an approximate solution